A27-xds-global-load-balancing.md

```diff
@@ -0,0 +1,426 @@
+----
+* Author(s): Mark D. Roth
+* Approver: a11r
+* Status: Draft
+* Implemented in: (in progress in C-core, Java, and Go)
+* Last updated: 2020-01-07
+* Discussion at: https://groups.google.com/d/topic/grpc-io/GWIH4XhIqHA/discussion
+
+## Abstract
+
+gRPC currently supports its own ["grpclb"
+protocol](https://github.com/grpc/grpc-proto/blob/master/grpc/lb/v1/load_balancer.proto)
+for [look-aside
+load-balancing](https://github.com/grpc/grpc/blob/master/doc/load-balancing.md).
+However, the popular [Envoy](http://envoyproxy.io) proxy uses the [xDS
+API](https://github.com/envoyproxy/data-plane-api/tree/master/envoy/api/v2)
+for many types of configuration, including load balancing, and that API
+is evolving into a standard that will be used to configure a variety of
+data plane software.  In order to converge with this industry trend, gRPC
+will be moving from its original grpclb protocol to the new xDS protocol.
+
+The xDS protocol allows configuring a variety of features, including
+many that gRPC does not currently support.  Over time, we will be slowly
+adding support for more of these features in gRPC.  We will publish a
+gRFC for each set of xDS features as we add support for them.
+
+This document describes the changes we are making to support global load
+balancing features, which is the set of xDS functionality that we are
+initially targetting.  This affects the gRPC client only; use of xDS in
+the gRPC server will come later as part of a separate set of xDS features.
+
+## Background
+
+The xDS API is actually a suite of APIs with names of the form "x
+Discovery Service", where there are many values for "x" (hence the name
+"xDS" for the overall protocol suite).  There is not currently a formal
+spec for the protocol, but the best reference is at
+https://www.envoyproxy.io/docs/envoy/latest/api-docs/xds_protocol.
+Readers of this document should familiarize themselves with the xDS API
+before proceeding.
+
+The xDS APIs are essentially pub/sub mechanisms, where each API allows
+subscribing to a different type of configuration resource.  In the xDS
+API flow, the client uses the following main APIs, in this order:
+
+- __Listener Discovery Service (LDS)__: Returns `Listener` resources.
+  Used basically as a convenient root for the gRPC client's configuration.
+  Points to the RouteConfiguration.
+- __Route Discovery Service (RDS)__: Returns `RouteConfiguration`
+  resources.  Provides data used to populate the gRPC service config.  Points
+  to the Cluster.
+- __Cluster Discovery Service (CDS)__: Returns `Cluster` resources.
+  Configures things like load balancing policy and load reporting.  Points
+  to the ClusterLoadAssignment.
+- __Endpoint Discovery Service (EDS)__: Returns `ClusterLoadAssignment`
+  resources.  Configures the set of endpoints (backend servers) to load
+  balance across and may tell the client to drop requests.
+
+gRPC will support the Aggregate Discovery Service (ADS) variant of xDS,
```

Comment:
Willl gRPC support both State of the World (SotW) and incremental variant of xDS or just one of them? 
Also does this mean that gRPC won't support basic xDS via gRPC and RESTful xDS?

Reply:
Initially, we will support only SotW ADS.  In the future, we may add support for (or maybe even switch to?) incremental ADS, since the incremental protocol's semantics are much clearer than the non-incremental protocol.  (We've run into a lot of edge cases while implementing this, which led to efforts like envoyproxy/envoy#9029 to try to clarify the exact semantics of the protocol.)

We do not currently plan to support non-aggregated xDS.  My understanding is that the trend right now is toward the aggregated variants of the protocol, and the current proposal for UDPA (the future evolution of xDS) is to offer only an aggregated protocol.  So we'd prefer not to put effort into older variants that aren't really going to be used in the future.

We do not plan to support REST.  After all, gRPC is intended to replace REST. :)

Reply:
Those all sounds reasonable to me. Maybe it would be useful to put some of the info here? Like only plan to support SotW ADS now and may switch to incremental ADS.

Reply:
Done.